extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/CustomQueryMethodsAdder.java

```diff
@@ -218,6 +229,32 @@ public void add(ClassCreator classCreator, FieldDescriptor entityClassFieldDescr
                         }
                     }
 
+                    // Find the type of data used in the result
+                    // e.g. method.returnType() is List, but it can contain custom/non-entity elements
+                    Type resultType = verifyQueryResultType(method.returnType());
+                    DotName customResultTypeName = resultType.name();
+
+                    if (customResultTypeName.equals(entityClassInfo.name())
+                            || customResultTypeName.equals(DotNames.OBJECT)) {
+                        // Result is using standard entity or Object result type 
+                        customResultTypeName = null;
+                    } else {
+                        // The result is using a custom type.
+                        List<String> fieldNames = GenerationUtil.getFieldNames(queryString);
+                        if (fieldNames.isEmpty()) {
+                            throw new IllegalArgumentException(
+                                    "Custom Query annotations must specify field aliases when used with a custom type. "
+                                            + "Offending query string is \"" + queryString + "\" on method " + methodName
+                                            + " of Repository " + repositoryName);
+                        }
+
+                        // If the custom type is an interface, we may need to generate the implementation
+                        ClassInfo resultClassInfo = index.getClassByName(customResultTypeName);
+                        if (Modifier.isInterface(resultClassInfo.flags())) {
```

Comment:
What happens if it's not an interface?

In other places we do have code that can deal with this situation (by instantiating a JavaBean for example), but does Spring Data support that?

Reply:
I don't know that Spring supports it, I can test. I'll also see if the newInstance works, though I imagine we'll also miss the field ordering.

Reply:
I throw an exception here. I could check to ensure the fields exist, but the code that attempts to convert from one type to the other depends on a method in the generated impl.

Reply:
Looks good